core/src/upgrade/transfer.rs

```diff
@@ -123,6 +124,38 @@ pub async fn read_length_prefixed(
     Ok(buf)
 }
 
+const DEFAULT_READ_STEP_SIZE: u64 = 1024;
+
+/// Reads a message from the given socket.
+///
+/// The `max_size` parameter is the maximum size in bytes of the message that we accept. This is
+/// necessary in order to avoid DoS attacks where the remote sends us a message of several
+/// gigabytes.
+pub async fn read_to_end(
+    socket: &mut (impl AsyncRead + Unpin),
+    max_size: usize,
+) -> io::Result<Vec<u8>> {
+    let step_size = min(DEFAULT_READ_STEP_SIZE, u64::try_from(max_size).unwrap());
+    let mut res = Vec::with_capacity(usize::try_from(step_size).unwrap());
+
+    loop {
+        let len = res.len();
+        if len > max_size {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!("Received data size ({len} bytes) exceeds maximum ({max_size} bytes)"),
+            ));
+        }
+
+        let num_bytes = socket.take(step_size).read_to_end(&mut res).await?;
+        if num_bytes == 0 {
+            break;
+        }
+    }
+
+    Ok(res)
+}
```

Comment:
I'd like to not add any more items to the public API of `libp2p-core` please. Let's keep this in `libp2p-request-response` for now, thanks!